Specs/GCMV8.lean

```diff
@@ -0,0 +1,262 @@
+-- irreducible polynomial Q(x) = <| 1 + x^^121 + x^^126 + x^^127 + x^^128 |>.
+
+/-- Binary representation for
+    irreducible polynomial P(x) = <| 1 + x + x^^2 + x^^7 + x^^128 |>
+-/
+def irrepoly : BitVec 129 := 0x100000000000000000000000000000087#129
+
+/-- Binary representation for
+    irreducible polynomial Q(x) = <| 1 + x^^121 + x^^126 + x^^127 + x^^128 |>
+    https://link.springer.com/chapter/10.1007/978-3-031-34671-2_30
+-/
+def refpoly : BitVec 129 := 0x1C2000000000000000000000000000001#129
+
+/-- H_hat : twisted H
+  H_hat = x × H (mod Q(x))
+  See Remark 5 in paper
+    "A New Interpretation for the GHASH Authenticator of AES-GCM"
+-/
+protected def gcm_init_H (H : BitVec 128) : BitVec 128 :=
+  pmod (H ++ 0b0#1) refpoly (by omega)
+
+protected def gcm_polyval_mul (x : BitVec 128) (y : BitVec 128) : BitVec 256 :=
+  0b0#1 ++ pmult x y
+
+protected def gcm_polyval_red (x : BitVec 256) : BitVec 128 :=
+  reverse $ pmod (reverse x) irrepoly (by omega)
+
+/--
+  The following implementation of POLYVAL first does pmult of x and y,
+  then reverses the input to project it into the GHASH field,
+  then does reduction in GHASH field and finally reverses the result back.
+  We could do this because of the following two facts:
+  1. Proposition 1 from paper
+    "A New Interpretation for the GHASH Authenticator of AES-GCM"
+  2. Lemma: reverse (pmult x y) = pmult (reverse x) (reverse y)
+-/
+protected def gcm_polyval (x : BitVec 128) (y : BitVec 128) : BitVec 128 :=
+  GCMV8.gcm_polyval_red $ GCMV8.gcm_polyval_mul x y
+
+/-- GCMInitV8 specification:
+    H : [128] -- initial H input
+    output : [12][128] -- precomputed Htable values
+-/
+def GCMInitV8 (H : BitVec 128) : (List (BitVec 128)) :=
+  let H0 := GCMV8.gcm_init_H H
+  let H2 := GCMV8.gcm_polyval H0 H0
+  let H1 := ((hi H2) ^^^ (lo H2)) ++ ((hi H0) ^^^ (lo H0))
+  let H3 := GCMV8.gcm_polyval H0 H2
+  let H5 := GCMV8.gcm_polyval H0 H3
+  let H4 := ((hi H5) ^^^ (lo H5)) ++ ((hi H3) ^^^ (lo H3))
+  let H6 := GCMV8.gcm_polyval H0 H5
+  let H8 := GCMV8.gcm_polyval H0 H6
+  let H7 := ((hi H8) ^^^ (lo H8)) ++ ((hi H6) ^^^ (lo H6))
+  let H9 := GCMV8.gcm_polyval H0 H8
+  let H11 := GCMV8.gcm_polyval H0 H9
+  let H10 := ((hi H11) ^^^ (lo H11)) ++ ((hi H9) ^^^ (lo H9))
+  [H0, H1, H2, H3, H4, H5, H6, H7, H8, H9, H10, H11]
+
+example :  GCMInitV8 0x66e94bd4ef8a2c3b884cfa59ca342b2e#128 =
+  [ 0xcdd297a9df1458771099f4b39468565c#128,
+    0x62d81a7fe5da3296dd4b631a4b7c0e2b#128,
+    0x88d320376963120dea0b3a488cb9209b#128,
+    0x8695e702c322faf935c1a04f8bfb2395#128,
+    0xb2261b4d0cb1e020b354474d48d9d96c#128,
+    0x568bd97348bd9145e4adc23e440c7165#128,
+    0xf9151b1f632d10b47d845b630bb0a55d#128,
+    0xa674eba8f9d7f2508491407c689db5e9#128,
+    0xec87cfb0e19d1c4e4af32418184aee1e#128,
+    0x7d1998bcfc545474f109e6e0b31d1eee#128,
+    0x7498729da40cd2808c107e5c4f494a9a#128,
+    0xd0e417a05fe61ba4a47c653dfbeac924#128 ] := by native_decide
```

Comment:
Unfortunately, `decide` and `rfl` don't work here. `decide` does print out a message that can help diagnose the problem --- maybe we can look at this later.